### DISCUSSION TITLE
How to catch ModuleNotFoundError using Pytest

### DISCUSSION THREAD

**Discussion by xuyxu** [pytest-dev/pytest#8269] [Q&A] [answered]
Hi, I met a problem when using Pytest to improve the code coverage of my own project.

The problem is that my project will automatically check that whether a third-party package is successfully installed using the code snippet below:

```
try:
    model = __import__("some_package")
except ModuleNotFoundError:
    msg = "Cannot load some_package."
    raise ModuleNotFoundError(msg)
```

Then, I would like to use `pytest.raises(ModuleNotFoundError)` to test the case that `some_package` is not installed. However, since `some_package` will be used for other tests, it is incorporated in the python environment of my CI/CD.

So, my question is that how can I do something like temporarily removing some installed packages (here, `some_package`) when Pytest is running a specific test case?

I would appreciate it very much if someone could help me with this. Thanks!

**Comment by graingert** [comment]
for this I use two tox environments, one with all my setuptools extras installed and one with none of them

**Reply by xuyxu** [reply]
Thanks for your reply!

This solves my question, but I am still wondering that whether there is a better solution, because setting two CI means I have smaller quota on Github Action.

**Comment by The-Compiler** [comment, marked as answer]
You can monkeypatch the `__import__` builtin, though it requires a bit of care. Something like:

```python3
import builtins

import pytest

@pytest.fixture
def patch_import(monkeypatch):
    real_import = __import__

    def fake_import(name, *args, **kwargs):
        if name != 'some_package':
            return real_import(name, *args, **kwargs)
        raise ImportError

    monkeypatch.setattr(builtins, '__import__', fake_import)
```

**Comment by xuyxu** [comment]
Thanks a lot @The-Compiler! It looks nice ;-)
